gubernator/github_auth.py

```diff
@@ -0,0 +1,133 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import urllib
+
+from webapp2_extras import security
+
+from google.appengine.api import urlfetch
+
+import view_base
+
+
+class Endpoint(view_base.BaseHandler):
+    def github_client(self):
+        client = self.app.config['github_client']
+        if client is None:
+            self.abort(500)
+        return client['id'], client['secret']
+
+    def maybe_redirect(self, target):
+        '''
+        Redirect to a given URL if it's determined to be safe.
+        '''
+        if target.startswith('/pr'):
+            self.redirect(target)
+
+    def get(self, arg):
+        # Documentation here: https://developer.github.com/v3/oauth/
+
+        client_id, client_secret = self.github_client()
+
+        if arg.endswith('/done'):
+            target, done = arg[:-len('/done')], True
+        else:
+            target, done = arg, False
+
+        if not done:
+            # 1) Redirect users to request GitHub access
+            if self.session.get('user'):
+                # User already logged in, no need to continue.
+                self.maybe_redirect(target)
+                return
+
+            state = security.generate_random_string(entropy=128)
+            args = {
+                'client_id': client_id,
+                'redirect_uri': self.request.url + '/done',
+                'scope': '',  # Username only needs "public data" permission!
+                'state': state
+            }
+            self.session['gh_state'] = state
+            self.redirect('https://github.com/login/oauth/authorize?'
+                + urllib.urlencode(args))
+        else:
+            # 2) GitHub redirects back to your site
+            code = self.request.get('code')
+            state = self.request.get('state')
+            session_state = self.session.pop('gh_state', '')
+
+            if not state or not code:
+                self.abort(400)
+            if not security.compare_hashes(state, session_state):
+                # States must match to avoid CSRF.
+                # Full attack details here:
+                # http://homakov.blogspot.com/2012/07/saferweb-most-common-oauth2.html
+                self.abort(400)
+
+            # 3) Use the access token to access the API
+            params = {
+                'client_id': client_id,
+                'client_secret': client_secret,
+                'code': code,
+                'state': session_state,
+            }
+            resp = urlfetch.fetch(
+                'https://github.com/login/oauth/access_token',
+                payload=urllib.urlencode(params),
+                method='POST',
+                headers={'Accept': 'application/json'}
+            )
+            if resp.status_code != 200:
+                logging.error('failed to vend token: status %d',
```

Comment:
"arguments on the first line forbidden when not using vertical alignment"

https://www.python.org/dev/peps/pep-0008/#indentation